### DISCUSSION TITLE
Switching from BEAKL-15 layout to Engram 2.0

### DISCUSSION THREAD

**Discussion by sunaku** [binarybottle/engram-layout#15] [Show and tell]
Hi Arno,

It's been a week since I've switched to the Engram 2.0 keyboard layout, so I thought I would share [my qualitative experience](https://www.keybr.com/profile/151qsif):

![progress](https://user-images.githubusercontent.com/9863/118227894-2c5a0080-b43e-11eb-9916-0dddd0cdea7b.png)

I switched over from the [BEAKL-15](https://deskthority.net/wiki/BEAKL) keyboard layout, which has _the same left-hand vowel cluster_ as Engram 2.0 --- this commonality significantly reduced my transition time. :sunglasses:  In contrast, it took me a month to become comfortable with that vowel cluster when I previously switched from Dvorak to BEAKL-15 six months ago.  Before that, I typed in Dvorak for 16 years and Qwerty for 6 years.

Coming to Engram, I just love the placement of punctuation in the middle of the keyboard to eliminate lateral movement of the index fingers while simultaneously providing quick access to frequently used punctuation (comma and period feel heavenly). I haven't seen any layout do this; especially spilling over the remaining letters to the right hand pinky cluster --- it's pure genius! :ok_hand:

The layout is very comfortable and lends itself well to programming.  In particular, I'm a die-hard Vim user and it passes my test:
* J and K are adjacent (identical to Dvorak actually) --- ideal :100: 
* H and L are adjacent (albeit on the same finger) --- good :+1: 
* W is on right-hand strong ring finger --- good :+1: 
* B is on left-hand weak pinky finger --- ouch :face_with_head_bandage: 
(but ring finger can easily hop over to take the load off the pinky --- okay :+1:)
* Punctuation is easily accessible to either hand since it's in the middle --- ideal :100: 
* Shifted punctuation on number keys are intuitive (or otherwise mnemonic) --- great :+1: 

I want to thank you for sharing your genius with the world. :bow:  This layout is truly a masterpiece; a game changer! :star_struck: Moreover, its systematic and methodical invention (and generous explanation thereof) allows interested people to understand _how_ this layout has come to be (by simply rendering the Python notebook on their own system) and to perform further research & improvement.

Cheers.

**Comment by binarybottle** [comment]
@sunaku -- Thank you very much for your extremely kind words.  It means a lot to me to know that someone has benefited from this work.  I could not have hoped for a more heartfelt testimonial! Please do keep me updated about your progress.

**Reply by sunaku** [reply]
Hi Arno,

After another week of real-world use (but this time without as much dedicated typing practice), I'm happy to say that I'm now comfortably typing at ~60 WPM as the new Engram muscle memory is finally taking precedence in my brain (thank goodness for neuroplasticity!). :nerd_face: 

I made a few tweaks to the orthonormal version of Engram on [my Ergodox EZ keymap](https://configure.zsa.io/ergodox-ez/layouts/dam5g/latest/0) for programming convenience and general comfort:
* I relocated the `@ # /` keys to the unused left-hand pinky keys (the slash is especially handy for URLs, file paths, and Vim searching)
* I put redundant square & angle brackets on either single-hand (otherwise the orthonormal split disconnects them, very far apart)
* I use [a dedicated symbol layer](https://configure.zsa.io/ergodox-ez/layouts/dam5g/latest/4) that further minimizes reaching by clustering frequent sigils & delimiters along the home row/block.

In addition, I would like to emphasize that Engram feels very comfortable to type in.  With BEAKL-15, I often felt like I'm "climbing a wall" when typing (for example, the word "spring" has me reaching up and down a lot).  Whereas with Engram, I feel like I'm "spreading my wings" and gracefully taking flight.  It's a significant qualitative improvement.

That's all for now, cheers!

**Reply by binarybottle** [reply]
These are great additions!  I was focusing on keeping standard, staggered keyboards and orthonormal layouts similar, but I really like your placement of the #,@, and / -- I might adopt these myself!  
And thank you for the update -- I am so happy to hear that you are continuing to benefit from and enjoy this layout!

**Reply by sunaku** [reply]
Hi Arno,

It's been a month since I've switched to Engram 2.0 and I'm happy to report that I'm comfortably typing at 70+ WPM. :star_struck:  

In particular, I've noticed that my interkey typing speed is the fastest I've ever observed in my 23 years of typing! :runner::dash: Portions of words (especially those in rolls) form rapidly on my screen as I type, though entire words take relatively longer due to momentary pauses as my muscle memory context switches between different finger/hand placements.  This is no doubt the result of those interkey speed and finger strength matrices applied from your extensive research. :ok_hand: 

Overall, I feel that I'm now at 95% confidence while typing: subconsciously finding the right keys to press and positioning my fingers in anticipation of the next characters coming down the pipeline.  Productivity wise, I feel 50% _more_ efficient.

Thanks again for all your hard work. :bow:  Cheers.

**Comment by binarybottle** [comment]
Thank you for the update, @sunaku! I'm so happy to hear that you're continuing to progress in your typing efficiency. Have you noticed any difference in terms of comfort when you type?

**Reply by sunaku** [reply]
Absolutely!  The speed & efficiency came hand in hand with comfort, not at its expense.  For example, the word "highlight" is delightful to type in Engram whereas I had to mentally prepare myself and concentrate to type it quickly in BEAKL-15.

However, I do notice minor slowdowns on upward-traveling same-finger bigrams such as "je" in the word "object", but this is more influenced by the style of keyboard being used than Engram itself.  For instance, I notice this more on my Ergodox, which has larger & distantly placed keys than those on my laptop.  In general, Engram scores well with low [1.24% SFBs](https://www.reddit.com/r/KeyboardLayouts/comments/ncewbo/arnos_engram_keyboard_layout/gz3oflk?utm_source=share&utm_medium=web2x&context=3). :ok_hand: 

Overall, Engram feels very comfortable due to its generous allotment of inward rolls (that "you" and "th" feel amazing), heavier emphasis on home row and bottom rows, balanced distribution between vertical climbs & horizontal spreads. :100:

**Reply by formlessdao** [reply]
What are the improvements of Engram over Colemak Mod-DH and Halmak?

**Reply by binarybottle** [reply]
@formlessdao -- Each of these layouts had different design considerations. For the motivations, protocol and design process for Engram, along with comparisons against these other layouts, please see the documentation in the Jupyter notebooks (https://github.com/binarybottle/engram), or outdated information in the preprint (https://www.preprints.org/manuscript/202103.0287/v1).

**Reply by sunaku** [reply]
@formlessdao -- regarding your question:

> What are the improvements of Engram over Colemak Mod-DH and Halmak?

I would say the main advantage is that Engram was designed from a clean slate with support for columnar split/ortholinear keyboards in mind. :bulb: Moreover, it doesn't limit itself to the historical legacy and physical constraints of typewriter style row-staggered keyboards.

For example, Engram takes full advantage of the former style of columnar keyboards (where each finger can travel straight up and down along its own dedicated column) by eliminating lateral movement of the index fingers into the central columns for typing letters (A-Z).

In contrast, Colemak assumes the latter style of row-staggered keyboards because it retains the ZXCV sequence from QWERTY legacy for application & clipboard shortcuts (which become far less important when you have a fully programmable keyboard that lets you place such shortcuts on dedicated keys, layers, combos, macros, etc. anywhere you like) and also because Colemak's variations (AngleMod, DH-mod, etc.) balance & redistribute finger workloads in a way that might not provide much benefit on columnar keyboards.  For instance, consider the AngleMod: how would you use that "angle" on keyboards with curved 3D keywells (e.g. Glove80, Dactyl, Kinesis, Maltron) anyway? :face_with_spiral_eyes:

Similarly, [Halmak's AI was trained on a MacBook laptop's row-staggered keyboard](https://youtu.be/pZ40gmfDFfQ), so it literally *assumes that exact* physical key placement as well as its trainer's natural home row finger placement (see "fingers are not perfect" at timestamp 5:11).  Thus, it might not provide as much benefit for columnar split/ortho keyboards or for those who have a different natural home row placement than the AI's trainer. :thinking:

**Reply by binarybottle** [reply]
Thank you for your excellent response, @sunaku!

**Comment by sunaku** [comment]
Hi Arno,

I've reached a comfortable speed of 100+ WPM with Engram 2.0 after properly configuring my programmable keyboard with [custom firmware that disambiguates "home row mods"](https://sunaku.github.io/home-row-mods.html) and allows me to type more naturally.

![monkeytype_20230202-233041](https://user-images.githubusercontent.com/9863/216539379-9ef07891-d4f6-4e4e-82ec-ac3facaad18c.png)

Looking back at my progress reported in this discussion thread, I see that I had [already reached a speed of 70+ WPM after a month](https://github.com/binarybottle/engram/discussions/15#discussioncomment-838277) of switching to Engram 2.0.  So it was my keyboard's firmware that held me back until now!

Cheers.

**Reply by binarybottle** [reply]
Good sleuthing and kudos for contributing patches to the firmware -- Thanks for sharing!

**Comment by coder054** [comment]
Switching L and D will make this layout even more vim friendly (the same with mtgap)

**Reply by sunaku** [reply]
Great observation! :thinking: I can see how that would make `H` and `L` more comfortable for Vim usage, but it would unfortunately come at the cost of increasing Same Finger Bigrams (SFBs) in Engram from 1.504% up to a significant 1.955% according to [Oxey's Layout Playground](https://o-x-e-y.github.io/layouts/playground/index.html).

**Comment by kayasoze** [comment]
Can you compare with BEAKL? Why did you switch?

**Reply by sunaku** [reply]
Great question!  I was asked the same thing a few years ago by Graeme Geldenhuys, who first [introduced me to BEAKL](https://www.reddit.com/r/ErgoDoxEZ/comments/gsvpug/layout_of_the_month_beakl_15/) and the [legendary Miryoku system](https://github.com/manna-harbour/miryoku#readme) of layers and home row mods.  This was my reply to him:

>> I'm curious what made you move away from BEAKL-15? Was there anything in particular that you did not like or struggle with.
>
>With BEAKL-15, I ultimately felt that I didn't experience the comfort that was advertised: I often felt like "climbing up a wall" when typing (for example, the word "spring" has me reaching up and down vertically). Also, I ended up [forking my own variation](https://github.com/sunaku/beakl15p) after 3 months because the punctuation assignments weren't as convenient as I liked for programming. Eventually, I found the obscure [BEAKL-19bis layout](https://github.com/bclnr/kb-layout-evaluation/issues/4) which scored well on a comparison page and I was interested in learning more about its origins (how did the authors invent it? what principles guided them?) but *I was too late:* like an archaeologist pondering ancient civilizations, so too, I was left with more questions than I could investigate and was left demoralized. Even the BEAKL authors (Den and Ian) disappeared, along with the Shenafu forums where they discussed their creations, just a few months before I even heard about BEAKL. :(
>
>Thankfully, I found that Ian Douglas [now @iandoug on GitHub] was still active elsewhere and actually ran the keyboard-design.com website which I encountered while investigating BEAKL-19bis. On that site, he mentioned Arno Klein's research publication and the resulting Engram layout. This caught my attention! Here was a layout, methodically & systematically derived from exhaustive comparison of permutations, guided by ergonomics principles. Not some golden result hidden behind AI algorithms like Halmak, or optimizers like AnDW for BEAKL-15, or CarpalX and MTGAP etc. Here was something that anyone could run on their own system and follow along in the Python notebook. (Plus the fact that Engram ended up on the same left-hand vowel layout as AnDW for BEAKL-15 greatly increased my confidence in the Engram methodology!)
>
>> Also did you use the BEAKL-15 number row and symbol suggestion?
>
>No, I didn't use the BEAKL-15 number row because it's asymmetric: 40123 76598 (I naturally expected 87659 on right hand for symmetry). Moreover, as a programmer, I type numbers even less than prose, so rearranging the already intuitive native number row didn't seem like a worthy trade-off in my case.
>
>As for the BEAKL-15 symbol layer, I adapted parts of its design into [my own symbol layer](https://configure.zsa.io/ergodox-ez/layouts/vKBnl/KyLLp/4) which is more optimized for Vim. For example, I have `^()$` on the left hand home row of my Symbol layer which lets me quickly travel to BOL PrevSentence NextSentence EOL respectively. You can see how the BEAKL-15 symbol layer isn't so convenient for Vim, although I do appreciate how logically it's laid out.

Note that my symbol layer has since [evolved into this](https://github.com/binarybottle/engram/discussions/45#discussioncomment-6389210) in the present day.

Cheers.

**Reply by binarybottle** [reply]
Thank you for telling us about your personal journey, @sunaku!

**Reply by kayasoze** [reply]
In comparing BEAKL 43 to Engram, I see that BEAKL has q, x and z in the weakest corners, but Engram has them on slightly stronger fingers. I'm a bit suspicious of this, but, on the other hand, BEAKL has more same finger bigrams on strong fingers (particularly around S). I have a keyboard with palm layer shifts, so I don't really care about the symbol placement on the main layer. Ultimately, I think, my pinkies are too weak to suffer Engram compared to BEAKL. I'll take the SFBs.

**Comment by binarybottle** [comment]
Please see https://github.com/binarybottle/engram-2021/discussions/86#discussion-10126586 for an announcement of the new Engram approach/layouts/repo!
